front/components/agent_builder/capabilities/knowledge/AddIncludeDataSheet.tsx

```diff
@@ -0,0 +1,166 @@
+import { DescriptionSection } from "@app/components/agent_builder/capabilities/knowledge/shared/DescriptionSection";
+import {
+  getDataSourceConfigurations,
+  getTimeFrame,
+  hasDataSourceSelections,
+  isValidPage,
+} from "@app/components/agent_builder/capabilities/knowledge/shared/sheetUtils";
+import { TimeFrameSection } from "@app/components/agent_builder/capabilities/knowledge/shared/TimeFrameSection";
+import type {
+  AgentBuilderAction,
+  IncludeDataAgentBuilderAction,
+} from "@app/components/agent_builder/types";
+import type {
+  DataSourceViewSelectionConfigurations,
+  TimeFrame,
+} from "@app/types";
+
+const PAGE_IDS = {
+  DATA_SOURCE_SELECTION: "data-source-selection",
+  CONFIGURATION: "configuration",
+} as const;
+
+type PageId = (typeof PAGE_IDS)[keyof typeof PAGE_IDS];
+
+interface AddIncludeDataSheetProps {
+  onSave: (action: AgentBuilderAction) => void;
+  isOpen: boolean;
+  onClose: () => void;
+  action?: AgentBuilderAction;
+}
+
+export function AddIncludeDataSheet({
+  onSave,
+  isOpen,
+  onClose,
+  action,
+}: AddIncludeDataSheetProps) {
+  const [currentPageId, setCurrentPageId] = useState<PageId>(
+    PAGE_IDS.DATA_SOURCE_SELECTION
+  );
+  const [description, setDescription] = useState(action?.description ?? "");
+  const [dataSourceConfigurations, setDataSourceConfigurations] =
+    useState<DataSourceViewSelectionConfigurations>(() =>
+      getDataSourceConfigurations(action)
+    );
+  const [timeFrame, setTimeFrame] = useState<TimeFrame | null>(() =>
+    getTimeFrame(action)
+  );
+
+  useEffect(() => {
+    setDescription(action?.description ?? "");
+    setDataSourceConfigurations(getDataSourceConfigurations(action));
+    setTimeFrame(getTimeFrame(action));
+  }, [action]);
+
+  const handleClose = useCallback(() => {
+    onClose();
+    setDescription("");
+    setDataSourceConfigurations({});
+    setTimeFrame(null);
+    setCurrentPageId(PAGE_IDS.DATA_SOURCE_SELECTION);
+  }, [onClose]);
+
+  const hasDataSources = useMemo(() => {
+    return hasDataSourceSelections(dataSourceConfigurations);
+  }, [dataSourceConfigurations]);
+
+  const handleSave = useCallback(() => {
+    const includeDataAction: IncludeDataAgentBuilderAction = {
+      id: action?.id || `include_data_${Date.now()}`,
+      type: "INCLUDE_DATA",
+      name: "Include Data",
+      description,
+      configuration: {
+        type: "INCLUDE_DATA",
+        dataSourceConfigurations,
+        timeFrame,
+      },
+      noConfigurationRequired: false,
+    };
+    onSave(includeDataAction);
+    handleClose();
+  }, [
+    action?.id,
+    description,
+    dataSourceConfigurations,
+    timeFrame,
+    onSave,
+    handleClose,
+  ]);
+
+  const handlePageChange = useCallback((pageId: string) => {
+    if (isValidPage(pageId, PAGE_IDS)) {
+      setCurrentPageId(pageId);
+    }
+  }, []);
+
+  const dataSourcePage = DataSourceSelectionPage({
```

Comment:
I'm not sure what is happening here, this is not returning a component? 

and useMemo below won't work because `dataSourcePage` will be re-created on every re-render right?

Reply:
We probably need to re-think how we handle pages, right now we re-render everything on state change like description 🙃 Let's remove useMemo, and stop using this weird `DataSourceSelectionPage` thingy, and revisit it later?

Reply:
Hmmm not sure what's so weird about this, but fine ditching it. The intent behind it was that it's the exact same component used in search and include.